# Part 1 - Python Documentation and Debugging Basics
Adapted the parts I found relevant from [GitHub - jakevdp/PythonDataScienceHandbook: Python Data Science Handbook: full text in Jupyter Notebooks](https://github.com/jakevdp/PythonDataScienceHandbook)

## 1. Documentation, matching and autocompletion
## 1.1 Documentation  - `?` or `help(Function)`

In [26]:
help(len)

Help on built-in function len in module builtins:

len(obj, /)
    Return the number of items in a container.



In [27]:
# Will display on panel at the bottom on Jupyter Notebook
len?

### Adding documentation to your own functions. - ` """ """ `

In [28]:
def square(a):
    """Return the square of a."""
    return a ** 2

In [29]:
help(square)

Help on function square in module __main__:

square(a)
    Return the square of a.



### Accessing Source Code - `??`
For functions implemented in Python, you can access source code. If you can't access source code, it is implemented in another language, e.g. C.

In [30]:
# Will display on panel at the bottom on Jupyter Notebook
square??

If function is implemented in another language, then `??` will return same thing as `?`

In [31]:
len??

## 1.2 Autocompletion - `[].<TAB>`

In [7]:
L = [1,2,3]

In [8]:
# Try L._<TAB>


Try `L._<TAB>`, you should options like:
    - L.append,
    - L.clear
    - L.copy
    ...

## 1.3 Wildcard Matching - `*[]?` or `[]*?` or `*[]*?`
Find objects in namespaces ( definition of namespace - https://code.tutsplus.com/tutorials/what-are-python-namespaces-and-why-are-they-needed--cms-28598 )

E.g. search for different types of warnings that are part of default python namespace, i.e. Search for objects ending with `Warning`

In [9]:
*Warning?

Search for objects part of numpy namespace, i.e. Search for objects beginning with `np.`

In [10]:
import numpy as np
np.*?

Search for any object that contains the word int anywhere. 

In [11]:
*int*?

******
## 2. Timing and Profiling

## 2.1 Timing - `timeit` or `time`

### 2.1.1 Timeit - `%%timeit` (multiline) or `%timeit` (singleline)
Runs multiple loops of lines of code and gives general timing info - best and worst runs.

In [32]:
%%timeit
L = []
for n in range(1000):
  L.append(n ** 2)  

240 µs ± 16.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


### 2.1.2 Time - `%time`
Timeit repeats actions to get average time and this could be misleading - e.g. when sorting a list, sorting a sorted list is faster then sorting an unsorted list.

In [33]:
import random
L = [random.random() for i in range(100000)]
print("sorting an unsorted list:")
%time L.sort()

sorting an unsorted list:
CPU times: user 28.5 ms, sys: 0 ns, total: 28.5 ms
Wall time: 28.4 ms


## 2.2 Profiling
Profiling can be useful to find what is causing a bottleneck in an application execution. 

You can profile each function call or you can profile line by line.

### Profiling Function Calls - `%prun [Function]`

In [34]:
def sum_of_lists(N):
    total = 0
    for i in range(5):
        L = [j ^ (j >> i) for j in range(N)]
        total += sum(L)
    return total

In [35]:
%prun sum_of_lists(1000000)

******
## 3. IPython's In and Out Objects
You can access previous inputs and outputs, using the `In` and `Out` objects.

In [36]:
import math

math.sin(2)

0.9092974268256817

In [37]:
math.cos(2)

-0.4161468365471424

### Print all inputs to iPython shell.

In [38]:
print(In)

['', 'help(len)', "# Will display on panel at the bottom on Jupyter Notebook\nget_ipython().run_line_magic('pinfo', 'len')", 'def square(a):\n    """Return the square of a."""\n    return a ** 2', 'help(square)', "# Will display on panel at the bottom on Jupyter Notebook\nget_ipython().run_line_magic('pinfo2', 'square')", "get_ipython().run_line_magic('pinfo2', 'len')", 'L = [1,2,3]', '# Try L._<TAB>', "get_ipython().run_line_magic('psearch', '*Warning')", "import numpy as np\nget_ipython().run_line_magic('psearch', 'np.*')", "get_ipython().run_line_magic('psearch', '*int*')", "get_ipython().run_cell_magic('timeit', '', 'L = []\\nfor n in range(1000):\\n  L.append(n ** 2)  ')", 'import random\nL = [random.random() for i in range(100000)]\nprint("sorting an unsorted list:")\nget_ipython().run_line_magic(\'time\', \'L.sort()\')', 'def sum_of_lists(N):\n    total = 0\n    for i in range(5):\n        L = [j ^ (j >> i) for j in range(N)]\n        total += sum(L)\n    return total', "get_ipy

### Print all iPython shell outputs

In [39]:
Out

{16: 0.9092974268256817,
 17: -0.4161468365471424,
 20: 'import math\n\nmath.sin(2)',
 36: 0.9092974268256817,
 37: -0.4161468365471424}

####  Format - `{a: b}` , where a is the input number that resulted in output b. 
E.g. `{16: 0.9092974268256817}`, meaning `In[16]` resulted in output `0.909...`

### Supressing output - ```;``` 
Add ```;``` to end of line, no output.

****************
## 4. Exceptions and Debugging
### 4.1 Exceptions
Different exceptions modes, with different details of information :  ```Plain```,```Context```and  ```Verbose```.

In [41]:
def func1(a, b):
    return a / b

def func2(x):
    a = x
    b = x - 1
    return func1(a, b)

Default xmode is `Context`

In [42]:
func2(1)

ZeroDivisionError: division by zero

In [43]:
%xmode Plain
func2(1)

Exception reporting mode: Plain


ZeroDivisionError: division by zero

In [44]:
%xmode Verbose
func2(1)

Exception reporting mode: Verbose


ZeroDivisionError: division by zero

### 4.2 Debugging - `%debug`
After an exception, if you run `%debug`, the debugger will open in interactive mode at the point of exception. You can then run commands and output variables and see what caused the exception.

In [45]:
%debug

> <ipython-input-41-586ccabd0db3>(2)func1()
      1 def func1(a, b):
----> 2     return a / b
      3 
      4 def func2(x):
      5     a = x

ipdb> exit


If you want to exit, while in interactive mode, type `exit`.